httpclient5/src/main/java/org/apache/hc/client5/http/entity/compress/Decoder.java

```diff
@@ -0,0 +1,60 @@
+/*
+ * ====================================================================
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * ====================================================================
+ *
+ * This software consists of voluntary contributions made by many
+ * individuals on behalf of the Apache Software Foundation.  For more
+ * information on the Apache Software Foundation, please see
+ * <http://www.apache.org/>.
+ *
+ */
+
+package org.apache.hc.client5.http.entity.compress;
+
+import java.io.IOException;
+
+import org.apache.hc.core5.http.HttpEntity;
+
+/**
+ * Pull-side transformer that takes a compressed {@link HttpEntity} and
+ * returns a lazily-decoded view of the same byte sequence.
+ *
+ * <p>Implementations <strong>must</strong> return an entity whose
+ * {@code InputStream} honours the usual stream contract and propagates
+ * {@link IOException}s raised by the underlying transport.</p>
+ *
+ * @since 5.6
+ */
+@FunctionalInterface
+public interface Decoder {
+
+    /**
+     * Wraps the supplied source entity in a decoding entity that transparently
+     * produces the uncompressed data.
+     *
+     * @param src the source {@code HttpEntity} positioned at the start of the
+     *            encoded payload (never {@code null})
+     * @return a new, undecoded {@code HttpEntity}; the caller is responsible
+     *         for consuming / closing its content stream
+     * @throws IOException if the decoding entity cannot be created or an
+     *                     underlying I/O error occurs
+     */
+    HttpEntity wrap(HttpEntity src) throws IOException;
```

Comment:
@arturobernalg I think this method does not need to throw `IOException`. At least it does not appear to be thrown anywhere in the existing code.

I also agree with @garydgregory here. I think we could just use standard `Function` here, but this is a matter of taste so I will not object.

Looks good otherwise.

Reply:
HI @ok2c  please do another pass. Now I'm using java functions instead.

Reply:
> HI @ok2c please do another pass. Now I'm using java functions instead.

@arturobernalg Oh, man. I meant something different. I think `IOFunction` was fine. It makes sense to have a function that throws IOException. What I meant the `Decoder` and `Encoder` could be just `UnaryOperator` or `Function` as suggested by @garydgregory